### DISCUSSION TITLE
When could be C#er could be  compiled into native code like Kotlin Native

### DISCUSSION THREAD

**Discussion by sgf** [dotnet/csharplang#1084] [General]
the Kotlin Native progress  is very fast(quickly).
now its 0.4 version. support windows platform begin v0.3 
and ppl can Write some program by it  and,
 run it on some platform with out any burden。

but almost all .net(C#) native project Not activily or Dead.

Can we  also expect ?

**Comment by sharwell** [comment]
Both the .NET Framework on Windows and .NET Core use a single-tier JIT (see [RyuJIT Tutorial](https://github.com/dotnet/coreclr/blob/master/Documentation/botr/ryujit-tutorial.md) and dotnet/coreclr#4331), so technically the current code is almost always compiled to native code before executing. While this is of course not what you meant, it doesn't seem like an irrelevant answer.

Many teams, both at Microsoft and external, have been working for a long time on improving the performance of .NET (including cases where C# code is the source) for real-world applications. Simply changing the point where C# code is converted to machine code will not necessarily improve an application. To give a better answer to your question, it would help greatly to understand the type of application involved and the reason why the current approaches are resulting in less-than-desired performance.

**Comment by Joe4evr** [comment]
You may be interested in [dotnet/corert](https://github.com/dotnet/corert).
> This repo contains CoreRT, a .NET Core runtime optimized for AOT (ahead of time compilation) scenarios, with the accompanying .NET Native compiler toolchain.

**Comment by benaadams** [comment]
Also crossgen https://github.com/dotnet/coreclr/blob/master/Documentation/building/crossgen.md

Doesn't make small native exes/dlls though

**Comment by MihaMarkic** [comment]
What's the difference between the two tools mentioned? Their goal looks similar at first glance.

**Comment by JoergWMittag** [comment]
Any language can be compiled to any other language, provided the target language is at least as computationally powerful as the source language. Since most modern mainstream CPU ISAs are Turing-complete, this means that any language can be natively compiled, including C♯.

> When could be C#er could be compiled into native code like Kotlin Native

It could be compiled into native code right now, and it always has and always will.

>Can we also expect ?

Sure. The only thing that we need is that someone needs to write a compiler that does that. The reason why we don't have a native implementation of C♯ yet is because there are too many people asking why we don't have a native implementation of C♯ yet and too few people writing a native implementation of C♯.
